File: AbstractFactory.cpp
#include<iostream>
using namespace std;
/*
* Abstract Factory
* What do client wants?
*		He wants to get enemies of a specified color without knowing how it works.

* How do we do?
*		We send a factory object to client, and he gets EVERYTHING!

* What's the advantage of this pattern?
*		We seperate products and client.

* What's the disadvantage of this pattern?
*	    We can hardly add a new color or a new kind of animal.
*/

//Abstract Product
class Bird {
public:
	virtual void attack() = 0;
};

//Concrete Product
class YellowBird : public Bird {
public:
	void attack() {
		cout << "Yellow Chew!" << endl;
	}
};

//Concrete Product
class GreenBird : public Bird {
public:
	void attack() {
		cout << "Green Chew!" << endl;
	}
};


class Cat {
	virtual void attack() = 0;
};

//Concrete Product
class YellowCat : public Cat {
public:
	void attack() {
		cout << "Yellow Meow!" << endl;
	}
};

//Concrete Product
class GreenCat : public Cat {
public:
	void attack() {
		cout << "Green Meow!" << endl;
	}
};


//Abstract Factory
class AbstractFactory {
public :
	virtual Bird* makeBird() = 0;
	virtual Cat* makeCat()  = 0;
};

//Concrete Factory
class YellowFactory : public AbstractFactory{
	Bird* makeBird() {
		return new YellowBird();
	}
	Cat* makeCat() {
		return new YellowCat();
	}
};
//Concrete Factory
class GreenFactory : public AbstractFactory {
	Bird* makeBird() {
		return new GreenBird();
	}
	Cat* makeCat() {
		return new GreenCat();
	}
};

//Client call this
void CreatedAndAttack(AbstractFactory* factory) {
	Bird* b = factory->makeBird();
	b->attack();

	Cat* c = factory->makeCat();
	b->attack();
}
int main() {
	YellowFactory* f = new YellowFactory();
	CreatedAndAttack(f);
	//Meow！

	cout << "-----------------------------" << endl;
	GreenFactory* f1 = new GreenFactory();
	CreatedAndAttack(f1);
	//Chew！
}

File: AbstractFactory/AbstractFactory.cs
using System;

namespace AbstractFactory
{
    interface IBird
    {
        abstract void attack();
    }

    class YellowBird : IBird
    {
        public void attack()
        {
            Console.WriteLine("Yellow Chew!");
        }
    }

    class GreenBird : IBird
    {
        public void attack()
        {
            Console.WriteLine("Green Chew!");
        }
    }


    interface ICat
    {
        abstract void attack();
    }

    class YellowCat : ICat
    {
        public void attack()
        {
            Console.WriteLine("Yellow Meow!");
        }
    }

    class GreenCat : ICat
    {
        public void attack()
        {
            Console.WriteLine("Green Meow!");
        }
    }

    interface IFactory
    {
        IBird makeBird();
        ICat makeCat();
    }

    class YellowFactory : IFactory
    {
        public IBird makeBird()
        {
            return new YellowBird();
        }
        public ICat makeCat()
        {
            return new YellowCat();
        }
    }

    class GreenFactory : IFactory
    {
        public IBird makeBird()
        {
            return new GreenBird();
        }
        public ICat makeCat()
        {
            return new GreenCat();
        }
    }

    class Client
    {
        //Client call this
        static void CreatedAndAttack(IFactory factory)
        {
            IBird b = factory.makeBird();
            b.attack();

            ICat c = factory.makeCat();
            c.attack();
        }
        static void Main()
        {
            IFactory f = new YellowFactory();
            CreatedAndAttack(f);
            //Meow！

            Console.WriteLine("-----------------------------");
            IFactory f1 = new GreenFactory();
            CreatedAndAttack(f1);
            //Chew！
        }
    }
}